pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+	return &fzw
+}
+
+func (fzw *FlatZincModel) AddParam(numInst uint, name, vartype, assignment string) {
+	fzw.ParamMap[name] = FlatZincParam{numInst, name, vartype, assignment}
+}
+
+func (fzw *FlatZincModel) AddVariable(numInst uint, name, vartype, assignment string, annotations ...string) {
+	fzw.VarMap[name] = FlatZincVariable{numInst, name, vartype, assignment, annotations}
+}
+
+func (fzw *FlatZincModel) AddConstraint(identifier string, exprs []string, annotations ...string) {
+	fzw.Constraints = append(fzw.Constraints, FlatZincConstraint{identifier, exprs, annotations})
+}
+
+func (fzw *FlatZincModel) SetSolveTarget(goal SolveGoal, expr string, annotations ...string) {
+	fzw.SolveTarget = FlatZincSolveItem{goal, expr, annotations}
+}
+
+func (fzw *FlatZincModel) Dump(fileName string) error {
+	file, err := os.Create(path.Clean(fileName))
+	if err != nil {
+		return fmt.Errorf("failed opening file %s for writing", fileName)
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Printf("Error closing file: %s\n", err)
+		}
+	}()
+
+	fileContent := ""
+	for _, fzparam := range fzw.ParamMap {
+		fileContent += fzparam.ParamDeclaration()
+	}
+
+	fileContent += "\n"
+	for _, fzvar := range fzw.VarMap {
+		fileContent += fzvar.VarDeclaration()
+	}
+
+	fileContent += "\n"
+	for _, cnstr := range fzw.Constraints {
+		fileContent += cnstr.ConstraintStatement()
+	}
+
+	fileContent += "\n" + fzw.SolveTarget.SolveItemStatement()
+	if _, err := file.WriteString(fileContent); err != nil {
+		return err
+	}
+	return nil
+}
+
+func parseSolutionLine(line string, lineNum uint) (string, []string, error) {
+	err := fmt.Errorf("parse error on line %d: %s", lineNum, line)
+	lineParts := strings.Split(line, "=")
+	if len(lineParts) != 2 {
+		return "", nil, err
+	}
+	varName := lineParts[0]
+	value := strings.TrimSuffix(lineParts[1], ";")
+	values := []string{value}
+	if strings.HasPrefix(value, "array") {
+		leftBracketPos := strings.Index(value, "[")
+		rightBracketPos := strings.Index(value, "]")
+		if leftBracketPos == -1 || rightBracketPos == -1 || leftBracketPos > rightBracketPos {
+			return "", nil, err
+		}
+		values = strings.Split(value[leftBracketPos+1:rightBracketPos], ",")
+	}
+	return varName, values, nil
+}
+
+func (fzw *FlatZincModel) ReadSolutions(fileName string) (map[string][]string, error) {
+	data, err := os.ReadFile(path.Clean(fileName))
+	if err != nil {
+		return nil, err
+	}
+	strContent := string(data)
+	lines := strings.Split(strContent, "\n")
+	newSolution := false
+	res := make(map[string][]string)
+	for lineNum, line := range lines {
+		line = strings.Join(strings.Fields(line), "") // remove all whitespaces
+		switch {
+		case len(line) == 0:
+			continue // empty line
+		case strings.HasPrefix(line, "%%%"):
+			continue // stat lines are ignored
+		case line == strings.Repeat("=", 10):
+			return res, nil // a solution was found and the whole search space was covered
+		case strings.HasPrefix(line, "===="):
```

Comment:
Does this mean UNSAT?  Wouldn't it be clearer to check for `UNSATISFIABLE` substring?

Reply:
There are several possible solver answers that have this `====` prefix, UNSAT being just one of them.
However, you are correct that we need to differentiate UNSAT from e.g., UNKNOWN, because an UNSAT may mean we need to retry with more modules.
I changed the code so that an UNSAT will return a slice with a single empty solution.
This way, legitimate solver outputs should always have one solution, and it's up to the caller to check that the solution is not empty.
If the result is not UNSAT and if no solution is found, we return with an error.